File: C++/2-Medium/00147._对链表进行插入排序/00147._对链表进行插入排序.cpp
#include <iostream>
#include <climits> // Include for INT_MIN
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

// 插入排序的实现
ListNode* insertionSortList(ListNode* head) {
    // 使用哑节点简化头部插入操作
    ListNode dummy(INT_MIN);
    ListNode *prev = &dummy;

    while (head != NULL) {
        ListNode *temp = head->next; // 保存下一个节点
        // 寻找插入位置
        if (prev->val >= head->val) prev = &dummy; // 重置prev为dummy，如果当前节点小于prev节点的值
        while (prev->next != NULL && prev->next->val < head->val) {
            prev = prev->next;
        }
        // 插入操作
        head->next = prev->next;
        prev->next = head;
        head = temp; // 移动到下一个待排序的节点
    }
    return dummy.next;
}

void printList(ListNode *head) {
    while (head != NULL) {
        cout << head->val << " ";
        head = head->next;
    }
    cout << endl;
}

int main() {
    ListNode *head = new ListNode(4);
    head->next = new ListNode(2);
    head->next->next = new ListNode(1);
    head->next->next->next = new ListNode(3);

    cout << "Original list: ";
    printList(head);

    ListNode *sortedList = insertionSortList(head);
    cout << "Sorted list: ";
    printList(sortedList);

    return 0;
}

File: Python/2-Medium/00147._对链表进行插入排序/00147._对链表进行插入排序.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def insertionSortList(head: Optional[ListNode]) -> Optional[ListNode]:
    """
    对链表进行插入排序。

    Args:
    head: ListNode - 链表的头节点。

    Returns:
    ListNode - 排序后的链表的头节点。
    """
    # 创建哑节点作为新链表的起始节点，便于处理边界情况
    dummy_head = ListNode(0)
    curr = head  # 当前正在排序的节点

    while curr:
        # prev 指针用于遍历找到 curr 应该插入的位置的前一个节点
        prev = dummy_head
        while prev.next and prev.next.val < curr.val:
            prev = prev.next

        # 将 curr 插入到 prev 和 prev.next 之间
        next_temp = curr.next  # 临时保存 curr 的下一个节点，因为插入过程中会改变 curr 的 next 指针
        curr.next = prev.next
        prev.next = curr
        curr = next_temp  # 移动 curr 到下一个待处理节点

    return dummy_head.next

# 辅助函数：创建链表
def create_list(arr):
    head = ListNode(arr[0]) if arr else None
    current = head
    for val in arr[1:]:
        current.next = ListNode(val)
        current = current.next
    return head

# 辅助函数：打印链表
def print_list(head):
    arr = []
    while head:
        arr.append(head.val)
        head = head.next
    return arr

# 示例链表 4->2->1->3
head = create_list([4, 2, 1, 3])
sorted_head = insertionSortList(head)

# 打印排序后的链表
print_list(sorted_head)